tests/functional/api/test_simple.py

```diff
@@ -31,3 +45,76 @@ def test_simple_api_detail(webtest):
     assert resp.content_type == "text/html"
     assert "X-PyPI-Last-Serial" in resp.headers
     assert f"Links for {project.normalized_name}" in resp.text
+
+
+def test_simple_attestations_from_upload(webtest):
+    user = UserFactory.create(
+        password=(  # 'password'
+            "$argon2id$v=19$m=1024,t=6,p=6$EiLE2Nsbo9S6N+acs/beGw$ccyZDCZstr1/+Y/1s3BVZ"
+            "HOJaqfBroT0JCieHug281c"
+        )
+    )
+    EmailFactory.create(user=user, verified=True)
+    project = ProjectFactory.create(name="sampleproject")
+    RoleFactory.create(user=user, project=project, role_name="Owner")
+    publisher = GitHubPublisherFactory.create(projects=[project])
+
+    # Construct the macaroon. This needs to be based on a Trusted Publisher, which is
+    # required to upload attestations
+    dm = MacaroonFactory.create(
+        oidc_publisher_id=publisher.id,
+        caveats=[
+            caveats.OIDCPublisher(oidc_publisher_id=str(publisher.id)),
+            caveats.ProjectID(project_ids=[str(p.id) for p in publisher.projects]),
+        ],
+        additional={"oidc": {"ref": "someref", "sha": "somesha"}},
+    )
+
+    m = pymacaroons.Macaroon(
+        location="localhost",
+        identifier=str(dm.id),
+        key=dm.key,
+        version=pymacaroons.MACAROON_V2,
+    )
+    for caveat in dm.caveats:
+        m.add_first_party_caveat(caveats.serialize(caveat))
+    serialized_macaroon = f"pypi-{m.serialize()}"
+
+    credentials = base64.b64encode(f"__token__:{serialized_macaroon}".encode()).decode(
+        "utf-8"
+    )
+
+    with open(_ASSETS / "sampleproject-3.0.0.tar.gz", "rb") as f:
+        content = f.read()
+
+    with open(
+        _ASSETS / "sampleproject-3.0.0.tar.gz.publish.attestation",
+    ) as f:
+        attestation = f.read()
+
+    expected_hash = hashlib.sha256(
+        # Filename:len(attestations)
+        b"sampleproject-3.0.0.tar.gz:1"
+    ).hexdigest()
+
+    webtest.post(
+        "/legacy/?:action=file_upload",
+        headers={"Authorization": f"Basic {credentials}"},
+        params={
+            "name": "sampleproject",
+            "sha256_digest": (
+                "117ed88e5db073bb92969a7545745fd977ee85b7019706dd256a64058f70963d"
+            ),
+            "filetype": "sdist",
+            "metadata_version": "2.1",
+            "version": "3.0.0",
+            "attestations": f"[{attestation}]",
+        },
+        upload_files=[("content", "sampleproject-3.0.0.tar.gz", content)],
+        status=HTTPStatus.OK,
+    )
+
+    response = webtest.get("/simple/sampleproject/", status=HTTPStatus.OK)
+    link = response.html.find("a", text="sampleproject-3.0.0.tar.gz")
+    assert "data-provenance" in link.attrs
+    assert link.get("data-provenance") == expected_hash
```

Comment:
This test fails here because the hash we get here is 

```python
hashlib.sha256(
    b"sampleproject-3.0.0.tar.gz:2"   # notice the 2 here
).hexdigest()
```

However, the `len(attestations)` should be 1.

The problem lies in [generate_provenance](https://github.com/pypi/warehouse/blob/e72b746d034b32f19c20eeefb11d5edf606d1011/warehouse/attestations/services.py#L141-L149) of `NullService` that creates the `DatabaseAttestation` and adds it to the `file.attestations` (like in [_persist_attestations](https://github.com/pypi/warehouse/blob/e72b746d034b32f19c20eeefb11d5edf606d1011/warehouse/attestations/services.py#L277-L287) )

Because the relationship is noted as [`back_populates`](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.back_populates), sqlalchemy automatically adds the newly created `DatabaseAttestation` to the `file` instance. 

The statement below is thus redundant and creates add a second time the attestation to the file instance.

```python
file.attestations.append(database_attestation)
```

However, if we look at [`test_persist_attestations_succeeds`](https://github.com/pypi/warehouse/blob/226204147a7c743b447a8b6014f955266b26824f/tests/unit/attestations/test_services.py#L86), the test passes with the following statements : 

```python
        assert len(attestations_db) == 1
        assert len(file.attestations) == 1
```

I've observed the state in a debugger using sqlalchemy tools and they looked appropriate. 
```
from sqlachemy import inspect
state = inspect(database_attestation)
```

Any idea here of the reason for this behavior ?

Reply:
The issue is that the `Attestation` objects weren't actually being added to the session, the test was emitting a warning like:

```
tests/unit/attestations/test_services.py::TestIntegrityService::test_generate_provenance_succeeds[GitHubPublisherFactory]
  /opt/warehouse/src/warehouse/attestations/models.py:53: SAWarning: Object of type <Attestation> not in session, add operation along 'File.attestations' will not proceed (This warning originated from the Session 'autoflush' process, which was invoked automatically in response to a user-initiated operation.)
    f"{Path(self.file.path).name}.attestation",
```

I think f7e277e and fa1bd58 should resolve this.

Reply:
Thanks for sorting this out - I missed the warning in the logs.